File: src/networking/java/net/remmintan/mods/minefortress/networking/c2s/ServerboundEditBlueprintPacket.java
package net.remmintan.mods.minefortress.networking.c2s;

import net.minecraft.block.BlockState;
import net.minecraft.network.PacketByteBuf;
import net.minecraft.server.MinecraftServer;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.server.world.ServerWorld;
import net.minecraft.util.BlockRotation;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Vec3i;
import net.remmintan.mods.minefortress.core.interfaces.blueprints.BlueprintDataLayer;
import net.remmintan.mods.minefortress.core.interfaces.blueprints.BlueprintGroup;
import net.remmintan.mods.minefortress.core.interfaces.blueprints.world.IBlueprintsWorld;
import net.remmintan.mods.minefortress.core.interfaces.entities.player.FortressServerPlayerEntity;
import net.remmintan.mods.minefortress.core.interfaces.networking.FortressC2SPacket;
import net.remmintan.mods.minefortress.core.interfaces.server.IFortressServer;

import java.util.HashMap;
import java.util.Map;

public class ServerboundEditBlueprintPacket implements FortressC2SPacket {

    private final Type type;
    private final String blueprintId;
    private final int floorLevel;
    private final BlueprintGroup blueprintGroup;

    private ServerboundEditBlueprintPacket(String blueprintId, int floorLevel, Type type, BlueprintGroup blueprintGroup) {
        this.blueprintId = blueprintId;
        this.floorLevel = floorLevel;
        this.type = type;
        this.blueprintGroup = blueprintGroup;
    }

    public ServerboundEditBlueprintPacket(PacketByteBuf buf) {
        this.blueprintId = buf.readString();
        this.floorLevel = buf.readInt();
        this.type = buf.readEnumConstant(Type.class);
        this.blueprintGroup = buf.readEnumConstant(BlueprintGroup.class);
    }

    @Override
    public void write(PacketByteBuf buf) {
        buf.writeString(this.blueprintId);
        buf.writeInt(this.floorLevel);
        buf.writeEnumConstant(type);
        buf.writeEnumConstant(blueprintGroup);
    }

    @Override
    public void handle(MinecraftServer server, ServerPlayerEntity player) {
        if(server instanceof IFortressServer fortressServer) {
            if(player instanceof FortressServerPlayerEntity fortressPlayer) {
                if(type == Type.REMOVE) {
                    fortressPlayer.get_ServerBlueprintManager().remove(blueprintId);
                } else {
                    final IBlueprintsWorld blueprintsWorld = fortressServer.get_BlueprintsWorld();
                    if(type == Type.EDIT) {
                        final var blockData = fortressPlayer
                                .get_ServerBlueprintManager()
                                .getBlockDataManager()
                                .getBlockData(blueprintId, BlockRotation.NONE);
                        final Map<BlockPos, BlockState> blueprintData = blockData
                                .getLayer(BlueprintDataLayer.GENERAL);

                        blueprintsWorld.prepareBlueprint(blueprintData, blueprintId, floorLevel, blueprintGroup);
                        blueprintsWorld.putBlueprintInAWorld(player, blockData.getSize());
                    } else if(type == Type.CREATE) {
                        blueprintsWorld.prepareBlueprint(new HashMap<>(), blueprintId, floorLevel, blueprintGroup);
                        blueprintsWorld.putBlueprintInAWorld(player, new Vec3i(1, 1, 1));
                    }
                    player.moveToWorld((ServerWorld) blueprintsWorld.getWorld());
                }
            }
        }
    }

    public static ServerboundEditBlueprintPacket edit(String name, int floorLevel, BlueprintGroup group) {
        return new ServerboundEditBlueprintPacket(name, floorLevel, Type.EDIT, group);
    }

    public static ServerboundEditBlueprintPacket add(String name, BlueprintGroup group) {
        return new ServerboundEditBlueprintPacket(name, 0, Type.CREATE, group);
    }

    public static ServerboundEditBlueprintPacket remove(String name) {
        return new ServerboundEditBlueprintPacket(name, 0, Type.REMOVE, BlueprintGroup.LIVING_HOUSES);
    }

    private enum Type {
        EDIT, CREATE, REMOVE
    }

}

File: src/networking/java/net/remmintan/mods/minefortress/networking/c2s/ServerboundEditBlueprintPacket.kt
package net.remmintan.mods.minefortress.networking.c2s

import net.minecraft.network.PacketByteBuf
import net.minecraft.server.MinecraftServer
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.util.BlockRotation
import net.minecraft.util.math.Vec3i
import net.remmintan.mods.minefortress.core.interfaces.blueprints.BlueprintDataLayer
import net.remmintan.mods.minefortress.core.interfaces.blueprints.BlueprintGroup
import net.remmintan.mods.minefortress.core.interfaces.blueprints.IStructureBlockData
import net.remmintan.mods.minefortress.core.interfaces.blueprints.ProfessionType
import net.remmintan.mods.minefortress.core.interfaces.networking.FortressC2SPacket
import net.remmintan.mods.minefortress.core.interfaces.server.IFortressServer
import net.remmintan.mods.minefortress.core.utils.getManagersProvider

class ServerboundEditBlueprintPacket : FortressC2SPacket {
    private val blueprintId: String?
    private val blueprintName: String?
    private val actionType: ActionType
    private val floorLevel: Int
    private val blueprintGroup: BlueprintGroup

    private constructor(
        blueprintId: String?,
        blueprintName: String?,
        floorLevel: Int,
        actionType: ActionType,
        blueprintGroup: BlueprintGroup
    ) {
        this.blueprintId = blueprintId
        this.blueprintName = blueprintName
        this.floorLevel = floorLevel
        this.actionType = actionType
        this.blueprintGroup = blueprintGroup
    }

    constructor(buf: PacketByteBuf) {
        this.blueprintId = buf.readNullable { obj: PacketByteBuf -> obj.readString() }
        this.blueprintName = buf.readNullable { obj: PacketByteBuf -> obj.readString() }
        this.floorLevel = buf.readInt()
        this.blueprintGroup = buf.readEnumConstant(BlueprintGroup::class.java)
        this.actionType = buf.readEnumConstant(
            ActionType::class.java
        )
    }

    override fun write(buf: PacketByteBuf) {
        buf.writeNullable(this.blueprintId) { b, id -> b.writeString(id) }
        buf.writeNullable(this.blueprintName) { b, name -> b.writeString(name) }
        buf.writeInt(this.floorLevel)
        buf.writeEnumConstant(blueprintGroup)
        buf.writeEnumConstant(actionType)
    }

    override fun handle(server: MinecraftServer, player: ServerPlayerEntity) {
        val managersProvider = getManagersProvider(player)
        if (player.server is IFortressServer) {
            val srv = player.server as IFortressServer
            val mp = player.getManagersProvider()
            if (actionType == ActionType.REMOVE) {
                mp.get_BlueprintManager().remove(blueprintId)
            } else {
                val blueprintsWorld = srv._BlueprintWorld
                if (actionType == ActionType.EDIT) {
                    val blockData: IStructureBlockData = mp.get_BlueprintManager()
                        .blockDataManager
                        .getBlockData(blueprintId, BlockRotation.NONE)
                    val blueprintData = blockData
                        .getLayer(BlueprintDataLayer.GENERAL)

                    val metadata = mp.get_BlueprintManager().get(blueprintId)

                    blueprintsWorld.setBlueprintMetadata(
                        player,
                        blueprintId,
                        blueprintName,
                        blueprintGroup,
                        metadata.capacity,
                        metadata.requirement.type ?: ProfessionType.NONE
                    )
                    blueprintsWorld.putBlueprintInAWorld(blueprintData, player, blockData.size, floorLevel)
                } else if (actionType == ActionType.CREATE) {
                    blueprintsWorld.setBlueprintMetadata(
                        player,
                        blueprintId,
                        blueprintName,
                        blueprintGroup,
                        4,
                        ProfessionType.NONE
                    )
                    blueprintsWorld.putBlueprintInAWorld(HashMap(), player, Vec3i(1, 1, 1), floorLevel)
                }
                player.moveToWorld(blueprintsWorld.world)
            }
        }
    }

    private enum class ActionType {
        EDIT, CREATE, REMOVE
    }

    companion object {
        @JvmStatic
        fun edit(id: String?, floorLevel: Int, group: BlueprintGroup): ServerboundEditBlueprintPacket {
            return ServerboundEditBlueprintPacket(id, null, floorLevel, ActionType.EDIT, group)
        }

        @JvmStatic
        fun add(name: String?, group: BlueprintGroup): ServerboundEditBlueprintPacket {
            return ServerboundEditBlueprintPacket(null, name, 0, ActionType.CREATE, group)
        }

        @JvmStatic
        fun remove(id: String?): ServerboundEditBlueprintPacket {
            return ServerboundEditBlueprintPacket(id, null, 0, ActionType.REMOVE, BlueprintGroup.LIVING_HOUSES)
        }
    }
}
